DOCS-5144 Adds Build an Integration, Build a Marketplace Offering, and Create a Tile Documentation (#14297)
DataDog/integrations-core: docs/dev/api_integration.md

---
title: Create an API Integration
type: documentation
further_reading:
- link: "/api/latest/using-the-api/"
  tag: "Documentation"
  text: "Learn how to use the Datadog API"
- link: "/developers/authorization/"
  tag: "Documentation"
  text: "Learn about using OAuth for API integrations"
- link: "/developers/"
  tag: "Documentation"
  text: "Learn how to develop on the Datadog platform"
---

## Overview

> This page walks Technology Partners through how to create a Datadog API integration on the [**Integrations** page][22]. 

Reviewer: ```suggestion
This page walks Technology Partners through how to create a Datadog API integration. 
```
It's rare, but you technically could create an api integration on the markeplace

---
title: Create an API Integration
type: documentation
further_reading:
- link: "/api/latest/using-the-api/"
  tag: "Documentation"
  text: "Learn how to use the Datadog API"
- link: "/developers/authorization/"
  tag: "Documentation"
  text: "Learn about using OAuth for API integrations"
- link: "/developers/"
  tag: "Documentation"
  text: "Learn how to develop on the Datadog platform"
---

## Overview

This page walks Technology Partners through how to create a Datadog API integration. 

## API integrations

Use [Datadog API endpoints][1] to enrich the customer's experience by submitting data from your backend and pulling data from a user's Datadog account. Technology Partners write and host their code within their environment. 

API integrations are ideal for Technology Partners that are SaaS-based, and have an existing platform that authenticates users.

API integrations can send the following types of data to Datadog:

- [Metrics][2]
- [Logs][3]
- [Events][4]
- [Service Checks][5]
- [Traces][6]
- [Incidents][7]

> You can also include out-of-the-box assets such as [monitors][25], [dashboards][26], and [log pipelines][27] with your integration. When a user clicks **Install** on your integration tile, they are prompted to follow the setup instructions, and all out-of-the-box dashboards and recommended monitors packaged with your integration will appear in the their account.

Reviewer: ```suggestion
You can include out-of-the-box assets such as [monitors][25], [dashboards][26], and [log pipelines][27] with your Agent-based integration. When a user clicks **Install** on your integration tile, they are prompted to follow the setup instructions, and all out-of-the-box dashboards will appear in their account. Other assets, such as log pipelines, will appear for users after proper installation and configuration of the integration.
```

> Since API integrations do not use the Datadog Agent to collect data, you need to create a tile (pictured below) to display your offering on the **Integrations** page once your development work is complete.

Reviewer 2: This line isn't accurate. All partner integrations require a tile, not just API integrations. 

I think the previous copy was clearer. For API integrations, the partner needs to add the steps to configure the integration in their tile.

## Overview

> Use [Datadog API endpoints][1] to enrich a customer's experience by submitting data from your backend and pulling data from a user's Datadog account. API integrations work well in building a connector between Datadog and another SaaS platform. Partners write and host their code within their environment.

Reviewer 3: ```suggestion
Use [Datadog API endpoints][1] to enrich a customer's experience by submitting data from your backend and pulling data from a user's Datadog account. API integrations work well when building a connector between Datadog and another SaaS platform. Partners write and host their code within their environment.
```
🤔

This method is ideal for Technology Partners that are SaaS based, and have an existing platform that authenticates users.

API integrations can send the following types of data to Datadog:

- [Metrics][2]
- [Logs][3]
- [Events][4]
- [Service Checks][5]
- [Traces][6]
- [Incidents][7]

> You can also include out-of-the-box assets like Monitors, Dashboards, and Log Pipelines with your integration. When a user clicks "Install" on your integration tile, they will be guided to your setup instructions, and all out-of-the-box dashboards and recommended monitors packaged with your integration will appear in the user's account.

Reviewer 3: Should this be called out in the agent docs too?

> This page provides instructions for creating an API integration. Since API integrations do not use the Datadog Agent to collect data, you need to create an informational tile (pictured below) to display your offering on the Integrations page once your development work is complete.

Reviewer 3: Perhaps "This page provides instructions for creating an API integration" could be moved to the top of the page.

{{< img src="developers/marketplace/marketplace-tile-example.png" alt="Example Marketplace tile" style="width:30%" >}}

> When a user clicks "Install" on your integration tile, they will be guided to your setup instructions, and all out-of-the-box dashboards and recommended monitors packaged with your integration will appear in the user's account.

Reviewer 3: This is already mentioned above.

## Setup

### Prerequisites

- You must implement OAuth in order to submit data or pull data out of Datadog. 

> - You must support all [Datadog sites][12].

Reviewer 3: Not sure if this is a prerequisite or more so part of the development process when implementing OAuth.

Reviewer: agree - removing this section entirely and mentioning sites in the oauth section

### OAuth
Instead of requesting API and Application keys directly from a user, Datadog requires using an [OAuth client][14] to handle authorization and access for API-based integrations. For more information, see [OAuth for Integrations][15] and [Authorization Endpoints][16]. 

You can explore examples that use OAuth in the `integrations-extras` repository such as [Vantage][17].

> ## Development Process

Reviewer 3: Should this be moved up to follow the same pattern as the agent docs?

You can expect the following process for building an API-based integration:
1. Meet with the Datadog Technology Partner team to discuss your integration and use cases.
2. Request a sandbox account for development.
3. Begin development of your integration, which will include writing and hosting integration code on your end, as well as implementing the OAuth protocol.
4. Test your integration, as well as your OAuth client, in your sandbox account.
5. Once your development work is tested and complete, follow the steps to [Create a Tile][24] in order to display your integration on the Marketplace or Integrations page.
6. Once your pull request is submitted and approved, the team will schedule a demo for a final review of your integration.
7. You'll have the option of testing the tile and integration in your sandbox account before publishing, or immediately publishing the integration for all customers. 

> ### To get started building a tile, [click here][24]. 

Reviewer 3: Is this supposed to be an H3? It threw me off cause none of the other docs have an H3 at the bottom.

Reviewer: We wanted it to be bold but agree it looks weird in the preview

## Overview

Use [Datadog API endpoints][1] to enrich a customer's experience by submitting data from your backend and pulling data from a user's Datadog account. API integrations work well in building a connector between Datadog and another SaaS platform. Partners write and host their code within their environment.

 This method is ideal for Technology Partners that are SaaS based, and have an existing platform that authenticates users.

API integrations can send the following types of data to Datadog:

- [Metrics][2]
- [Logs][3]
- [Events][4]
- [Service Checks][5]
- [Traces][6]
- [Incidents][7]

You can also include out-of-the-box assets like Monitors, Dashboards, and Log Pipelines. When a user clicks "Install" on your integration tile, they will be guided to your setup instructions, and all out-of-the-box dashboards and recommended monitors packaged with your integration will appear in the user's account.

> This page provides instructions for creating an API integration. Since API integrations do not use the Datadog Agent to collect data, you need to create an informational tile-only listing to display your offering on the Integrations page once your development work is complete.

Reviewer 4: I think this should be moved up with the sentence: 
 This method is ideal for Technology Partners that are SaaS based, and have an existing platform that authenticates users.

## Overview

Use [Datadog API endpoints][1] to enrich a customer's experience by submitting data from your backend and pulling data from a user's Datadog account. API integrations work well in building a connector between Datadog and another SaaS platform. Partners write and host their code within their environment.

 This method is ideal for Technology Partners that are SaaS based, and have an existing platform that authenticates users.

API integrations can send the following types of data to Datadog:

- [Metrics][2]

> - [Logs & Log Pipelines][3]

Reviewer 2: They also don't send service checks.

What's a security event? I don't believe there is any way to send a security event into Datadog. Instead, partners can send logs, which can be compared against security rules, but they are still sending logs. 

Log pipelines, dashboards, monitors, security rules, etc. are assets, though, not data types.

> - [Logs & Log Pipelines][3]

Reviewer 4: - [Metrics][2]
- [Logs][3]
- [Events][4]
- [Service Checks][5]
- [Traces][6]
- [Incidents][7]
- [Security Rules][8]

OR 

API integrations can send the following types of data to Datadog:
- [Metrics][2]
- [Logs][3]
- [Events][4]
- [Service Checks][5]
- [Traces][6]
- [Incidents][7]
- [Security Rules][8]

You can also include out-of-the-box assets like Monitors, Dashboards, and Security Rules. When a user clicks "Install" on your integration tile, they will be guided to your setup instructions, and all out-of-the-box dashboards and recommended monitors packaged with your integration will appear in the user's account.

> - [Logs & Log Pipelines][3]

Reviewer: I like the second one, but let's hold off on mentioning security rules as an asset for now as that's not built yet

> - [Logs & Log Pipelines][3]

Reviewer 4: mention log pipeline in assets

- [Incidents][7]
- [Security Events][8]

> This page provides instructions for creating an API integration. REMOVE: For more information about why you would want to create an API-based integration, see [Creating your own solution][9]. Since API integrations do not use the Datadog Agent to collect data, you need to create an informational tile-only listing once your development work is complete. 

Reviewer 4: The first two bullets are touched on in this page but added a sentence on write and host their own code: 

> Use a [Datadog API endpoint][1] to enrich and submit data from your backend. API integrations work well in building a connector between Datadog and another SaaS platform. Partners will write and host their code within their enviornment.
This method is ideal for Technology Partners that are SaaS based, and have an existing website for users to log into for authorization purposes.

> This page provides instructions for creating an API integration. REMOVE: For more information about why you would want to create an API-based integration, see [Creating your own solution][9]. Since API integrations do not use the Datadog Agent to collect data, you need to create an informational tile-only listing once your development work is complete. 

Reviewer 4: I am unable to find the page that lists all scopes - does this exist?

> This page provides instructions for creating an API integration. REMOVE: For more information about why you would want to create an API-based integration, see [Creating your own solution][9]. Since API integrations do not use the Datadog Agent to collect data, you need to create an informational tile-only listing once your development work is complete. 

Reviewer 4: We are already linking to a lot of articles on this page: 
### Create an OAuth client
Instead of requesting API and Application keys directly from a user, Datadog requires using an [OAuth client][14] to handle authorization and access for API-based integrations. For more information, see [OAuth for Integrations][15] and [Authorization Endpoints][16]. 

It may be too much linking to a list of scopes. Is there a way to condense this?

> This page provides instructions for creating an API integration. REMOVE: For more information about why you would want to create an API-based integration, see [Creating your own solution][9]. Since API integrations do not use the Datadog Agent to collect data, you need to create an informational tile-only listing once your development work is complete. 

Reviewer: I think it would be ideal if all the bullet points I mentioned above were in the overview section. For the OAuth piece, we can leave the scopes link out if it's too much - the oauth docs link to it anyways

- [Incidents][7]
- [Security Events][8]

This page provides instructions for creating an API integration. Since API integrations do not use the Datadog Agent to collect data, you need to create an informational tile-only listing to display your offering on the Integrations page once your development work is complete.

When a user clicks "Install" on your integration tile, they will be guided to your setup instructions, and all out-of-the-box dashboards and recommended monitors packaged with your integration will appear in the user's account.

## Setup

### Prerequisites

- You must implement OAuth in order to submit data or pull data out of Datadog. 
- You must support all [Datadog sites][12].

### OAuth
Instead of requesting API and Application keys directly from a user, Datadog requires using an [OAuth client][14] to handle authorization and access for API-based integrations. For more information, see [OAuth for Integrations][15] and [Authorization Endpoints][16]. 

You can explore examples that use OAuth in the `integrations-extras` repository such as [Vantage][17].

> ## Development Process

Reviewer 3: Was this included in the agent docs?

Reviewer 4: add same section to agent docs

You can expect the following process for building an API-based integration:
1. Meet with the Datadog Technology Partner team to discuss your offering and use cases.

> 2. Request a sandbox account for development.

Reviewer 2: What happened to the steps describing how to become a Technology Partner? 

Should those continue to be laid out up front? Otherwise, it's not clear how the user gets a sandbox account, and they will probably get delayed by emailing support or our team directly.

Reviewer 4: The steps are outlined in Build a Marketplace Offering page. Do you think they should be located somewhere else?

Reviewer 4: remove from marketplace offering page to readme.

3. Begin development of your integration, which will include writing and hosting integration code on your end, as well as implementing the OAuth protocol.
4. Test your integration, as well as your OAuth client, in your sandbox account.
5. Once your development work is tested and complete, **follow the steps to [Create a Tile][24] in order to display your offering on the Marketplace or Integrations page**.
6. Once your pull request is submitted and approved, the team will schedule a demo for a final review of your integration.
7. You'll have the option of testing the tile and integration in your sandbox account before publishing, or immediately publishing the integration for all customers. 

> To get started building a tile, [click here][24]. 

Reviewer 2: Just my two cents, but the bolding in number 5 with the redundant links are a tad confusing.

To get started, you can explore examples that use OAuth in the `integrations-extras` repository such as [Vantage][17].

### Build your integration

The process to build an API-based integration looks like the following:

1. Once you've been accepted to the Datadog Partner Network, you will meet with the Datadog Technology Partner team to discuss your offering and use cases.
2. Request a Datadog sandbox account for development.
3. Begin development of your integration, which includes writing and hosting the integration code on your end as well as implementing the [OAuth protocol][28].
4. Test your integration, as well as your OAuth client, in your Datadog sandbox account.
5. Once your development work is tested and complete, follow the steps in [Create a Tile][20] in order to display your integration on the **Integrations** or **Marketplace** page.
6. Once your pull request is submitted and approved, the Datadog Technology Partner team will schedule a demo for a final review of your integration.
7. You will have the option of testing the tile and integration in your Datadog sandbox account before publishing, or immediately publishing the integration for all customers.  

> To learn about creating a tile for your offering, see [Create a Tile][24].

Reviewer: Could we have something more actionable here? 
```suggestion
Get started building your API integration and [create a tile][24].
```

> 

Reviewer 4: @anawishnoff should we change: Determine which [Datadog site][12] you want to use. ---> You must support all Datadog Sites?

> 

Reviewer: Yeah that makes more sense, and we can hyperlink to the doc on sites: https://docs.datadoghq.com/getting_started/site/

You can explore examples of existing API integrations in the `integrations-extras` repository such as [Vantage][17].

> To create an API integration, click here.[24]

Reviewer 4: @alai97 Quick question - Do other docs put the process in the beginning or end of the doc?

> To create an API integration, click here.[24]

Author: I believe docs like the Mobile App (https://docs.datadoghq.com/mobile/?tab=ios) page have instructions for users to jump right into it!

[21]: https://github.com/DataDog/integrations-extras/
[22]: https://app.datadoghq.com/integrations
[23]: /developers/integrations/python

> [24]: Create a tile link

Author: ```suggestion
[24]: https://docs.datadoghq.com/developers/integrations/create_a_tile
```
